test/asynchronous/test_database.py

```diff
@@ -0,0 +1,778 @@
+    """Test Database features on a client that does not connect."""
+
+    client: AsyncMongoClient
+
+    @classmethod
+    def setUpClass(cls):
+        cls.client = AsyncMongoClient(connect=False)
+
+    def test_name(self):
+        self.assertRaises(TypeError, AsyncDatabase, self.client, 4)
+        self.assertRaises(InvalidName, AsyncDatabase, self.client, "my db")
+        self.assertRaises(InvalidName, AsyncDatabase, self.client, 'my"db')
+        self.assertRaises(InvalidName, AsyncDatabase, self.client, "my\x00db")
+        self.assertRaises(InvalidName, AsyncDatabase, self.client, "my\u0000db")
+        self.assertEqual("name", AsyncDatabase(self.client, "name").name)
+
+    def test_get_collection(self):
+        codec_options = CodecOptions(tz_aware=True)
+        write_concern = WriteConcern(w=2, j=True)
+        read_concern = ReadConcern("majority")
+        coll = self.client.pymongo_test.get_collection(
+            "foo", codec_options, ReadPreference.SECONDARY, write_concern, read_concern
+        )
+        self.assertEqual("foo", coll.name)
+        self.assertEqual(codec_options, coll.codec_options)
+        self.assertEqual(ReadPreference.SECONDARY, coll.read_preference)
+        self.assertEqual(write_concern, coll.write_concern)
+        self.assertEqual(read_concern, coll.read_concern)
+
+    def test_getattr(self):
+        db = self.client.pymongo_test
+        self.assertTrue(isinstance(db["_does_not_exist"], AsyncCollection))
+
+        with self.assertRaises(AttributeError) as context:
+            db._does_not_exist
+
+        # Message should be: "AttributeError: Database has no attribute
+        # '_does_not_exist'. To access the _does_not_exist collection,
+        # use database['_does_not_exist']".
+        self.assertIn("has no attribute '_does_not_exist'", str(context.exception))
+
+    def test_iteration(self):
+        db = self.client.pymongo_test
+        if "PyPy" in sys.version and sys.version_info < (3, 8, 15):
+            msg = "'NoneType' object is not callable"
+        else:
+            if _IS_SYNC:
+                msg = "'Database' object is not iterable"
+            else:
+                msg = "'AsyncDatabase' object is not iterable"
+        # Iteration fails
+        with self.assertRaisesRegex(TypeError, msg):
+            for _ in db:  # type: ignore[misc] # error: "None" not callable  [misc]
+                break
+        # Index fails
+        with self.assertRaises(TypeError):
+            _ = db[0]
+        # next fails
+        with self.assertRaisesRegex(TypeError, "'Database' object is not iterable"):
+            _ = next(db)
+        # .next() fails
+        with self.assertRaisesRegex(TypeError, "'Database' object is not iterable"):
+            _ = db.next()
+        # Do not implement typing.Iterable.
+        self.assertNotIsInstance(db, Iterable)
+
+
+class TestDatabase(AsyncIntegrationTest):
+    async def test_equality(self):
```

Comment:
Only methods that call `async` functions need to be asynchronous. In general if there aren't any `await`s used in a method it doesn't have to be asynchronous.